crates/bdk/src/wallet/mod.rs

```diff
@@ -2392,6 +2393,82 @@ fn create_signers<E: IntoWalletDescriptor>(
     Ok((signers, change_signers))
 }
 
+/// An implementation of [`BIP326`] Anti-fee sniping in Taproot transactions. This function
+/// sets the transaction [`LockTime`].
+///
+/// [`BIP326`]: (https://github.com/bitcoin/bips/blob/master/bip-0326.mediawiki)
+/// [`LockTime`]: (absolute::LockTime)
+fn create_tx_locktime(
+    transaction: &mut Transaction,
+    selected: &[Utxo],
+    chain_tip: u32,
+    rbf_enabled: bool,
+    policy_fixed: bool,
+) {
+    // nSequence can encode a max csv of 65535, detailed in BIP68
+    const MAX_ENCODABLE_CSV: u32 = 0xFFFF;
+    let mut rng = rand::thread_rng();
+
+    // Filter foreign utxos, since we don't know their confirmation time
+    let utxos: Vec<&LocalOutput> = selected
+        .iter()
+        .filter_map(|u| match u {
+            Utxo::Local(txo) => Some(txo),
+            _ => None,
+        })
+        .collect();
+
+    /* Set transaction `LockTime` to the current height if:
+    - 50% probability
+    - RBF not enabled
+    - any inputs:
+        - are not p2tr
+        - have more than 65535 confirmations
+        - are unconfirmed */
+    if rand::random()
+        || !rbf_enabled
+        || utxos.iter().any(|u| {
+            !u.txout.script_pubkey.is_v1_p2tr()
+                || match u.confirmation_time {
+                    ConfirmationTime::Confirmed { height, .. } => {
+                        chain_tip - height > MAX_ENCODABLE_CSV
+                    }
+                    _ => true, // unconfirmed
+                }
+        })
+    {
+        let mut height = chain_tip;
+        if rng.gen_range(0..10) == 1 {
+            // Occasionally, randomly pick a locktime further back.
+            height = height.saturating_sub(rng.gen_range(0u32..100));
+        }
+
+        transaction.lock_time =
+            absolute::LockTime::from_height(height).expect("valid locktime height");
+    } else {
+        // If none of the above are true, we're spending taproot coins with fewer
+        // than 65535 confirmations, and RBF enabled. Set `LockTime::ZERO`, and set
+        // the `Sequence` of a randomly chosen input to its confirmations.
+        if !policy_fixed && !utxos.is_empty() {
+            let idx = rng.gen_range(0..utxos.len());
+            let utxo = &utxos[idx];
+            let mut confs = match utxo.confirmation_time {
+                ConfirmationTime::Confirmed { height, .. } => {
+                    (chain_tip - height).try_into().expect("confs is valid u16")
+                }
+                _ => 0u16,
+            };
+            if rng.gen_range(0..10) == 1 {
+                // Occasionally, randomly pick a lower sequence.
+                confs = confs.saturating_sub(rng.gen_range(0u16..100));
+            }
+            transaction.input[idx].sequence = Sequence::from_height(confs);
```

Comment:
This needs to find the input corresponding to the randomly selected utxo above, rather than assume the same `idx`